src/libtools/graphalgorithm.cc

```diff
@@ -48,77 +49,265 @@ vector<Edge> edgeSetToVector_(set<Edge> edges) {
+    for (const int& vertex : remaining_unexplored) {
+      if (!vertex_explored_[vertex].first) return vertex;
+    }
+
+    throw runtime_error(
+        "You cannot get an unexplored vertex as they have all"
+        " been explored.");
+  }
+};
+
+ReducedGraph reduceGraph(Graph graph) {
+
+  unordered_map<int, pair<bool, int>> unexplored_vertices;
+  vector<int> vertices = graph.getVertices();
+  for (const int& vertex : vertices) {
+    unexplored_vertices[vertex] =
+        pair<bool, int>(false, graph.getDegree(vertex));
+  }
+
+  ExplorationRecord exploration_record(unexplored_vertices);
+
+  vector<vector<int>> chains;
+
+  while (exploration_record.unexploredVerticesExist()) {
+    Graph_DF_Visitor graph_visitor;
+    int starting_vertex = exploration_record.getUnexploredVertex();
+    exploration_record.explore(starting_vertex);
+    graph_visitor.setStartingVertex(starting_vertex);
+    graph_visitor.initialize(graph);
+
+    vector<int> chain{starting_vertex};
+    int old_vertex = starting_vertex;
+    bool new_chain = false;
+    while (!graph_visitor.queEmpty()) {
+      Edge edge = graph_visitor.nextEdge(graph);
+      vector<int> unexplored_vertex = graph_visitor.getUnexploredVertex(edge);
 
-  auto vertices_list = vectorToList_(g.getVertices());
+      if (new_chain) {
+        if (unexplored_vertex.size() == 0) {
+          old_vertex = edge.getEndPoint1();
+          chain.push_back(old_vertex);
+          new_chain = false;
+        } else {
+          old_vertex = edge.getOtherEndPoint(unexplored_vertex.at(0));
+          chain.push_back(old_vertex);
+          new_chain = false;
+        }
+      }
+      int new_vertex = edge.getOtherEndPoint(old_vertex);
+      if (unexplored_vertex.size() == 0) {
+        chain.push_back(new_vertex);
+        chains.push_back(chain);
+        chain.clear();
+        new_chain = true;
+      } else if (graph.getDegree(new_vertex) == 1) {
+        chain.push_back(new_vertex);
+        chains.push_back(chain);
+        chain.clear();
+        exploration_record.explore(new_vertex);
+        new_chain = true;
+      } else if (graph.getDegree(new_vertex) > 2) {
+        chain.push_back(new_vertex);
+        chains.push_back(chain);
+        chain.clear();
+        exploration_record.explore(new_vertex);
+        new_chain = true;
+      } else if (unexplored_vertex.size() == 1) {
+        chain.push_back(new_vertex);
+        old_vertex = new_vertex;
+        exploration_record.explore(new_vertex);
+      }
+
+      graph_visitor.exec(graph, edge);
+    }
+  }
+  vector<ReducedEdge> reduced_edges;
+  for (vector<int> chain : chains) {
+    ReducedEdge reduced_edge(chain);
+    reduced_edges.push_back(reduced_edge);
+  }
+
+  ReducedGraph reduced_g(reduced_edges);
+  auto nodes_graph = graph.getNodes();
+  reduced_g.copyNodes(graph);
+  auto nodes_reduced_g = reduced_g.getNodes();
+  return reduced_g;
+}
+
+vector<shared_ptr<Graph>> decoupleIsolatedSubGraphs(Graph graph) {
```

Comment:
why shared_ptr?

Reply:
This is left over from when I was confused about how memory was being handled. I will have to go back and fix this.

Reply:
shared_ptr was removed switched to returning a vector of graphs.